pkg/cloudprovider/openstack/openstack.go

```diff
@@ -625,28 +626,41 @@ func (lb *LoadBalancer) UpdateTCPLoadBalancer(name, region string, hosts []strin
 func (lb *LoadBalancer) DeleteTCPLoadBalancer(name, region string) error {
 	glog.V(4).Infof("DeleteTCPLoadBalancer(%v, %v)", name, region)
 
-	vip, err := getVipByName(lb.network, name)
-	if err != nil {
-		return err
+	vip, vipErr := getVipByName(lb.network, name)
+	vipExists := true
+	if vipErr != nil {
+		if vipErr != ErrNotFound {
+			return vipErr
+		}
+		vipExists = false
 	}
 
-	pool, err := pools.Get(lb.network, vip.PoolID).Extract()
-	if err != nil {
-		return err
+	pool, poolErr := pools.Get(lb.network, vip.PoolID).Extract()
+	poolExists := true
+	if poolErr != nil {
+		detailedErr, ok := poolErr.(*gophercloud.UnexpectedResponseCodeError)
+		if !ok || detailedErr.Actual != http.StatusNotFound {
+			return poolErr
+		}
+		poolExists = false
 	}
 
 	// Have to delete VIP before pool can be deleted
-	err = vips.Delete(lb.network, vip.ID).ExtractErr()
-	if err != nil {
-		return err
+	if vipExists {
+		err := vips.Delete(lb.network, vip.ID).ExtractErr()
+		if err != nil {
+			return err
+		}
 	}
 
 	// Ignore errors for everything following here
 
-	for _, monId := range pool.MonitorIDs {
-		pools.DisassociateMonitor(lb.network, pool.ID, monId)
+	if poolExists {
+		for _, monId := range pool.MonitorIDs {
+			pools.DisassociateMonitor(lb.network, pool.ID, monId)
+		}
+		pools.Delete(lb.network, pool.ID)
```

Comment:
We need to delete the vip before the pool (otherwise the pool is still in use), and once the vip is deleted GetTCPLoadBalancerExists will indicate "does not exist", and we could potentially recreate another vip with the same name (and a new pool, also with the same name).  My intention (and the current code) was just to leak the pool object during partial failures (the "ignore errors following here" comment could have been clearer).

Now that we're looking at retrying the delete operation, we should definitely disassociate (and delete!) the monitors before attempting to remove the vip (partial failures will lead to missing health checks, but if DeleteLoadBalancer is retried then eventually we might delete everything successfully).  If the disassociate succeeds and the monitor delete fails, we're still leaking the monitor.

The pool is still awkward.  It is possible to lookup the pool by name (rather than ID), but the name is not necessarily unique - if we have already recreated a new LoadBalancer with the same name then we could delete the wrong pool.  With a bit more clarity around how k8s will interpret "Get" and "Create" when "Delete" fails we could assume the pool name is unique and use that to retry deleting the pool.

Less desirable alternatives might include trying to "unwind" and recreate the vip if there were failures deleting the pool (yuck); using OpenStack metadata to associate pool+monitor to original LB (I don't modify metadata anywhere else); or modifying k8s to allow cloud providers to store resource metadata somewhere within k8s (bigger change, but could also lead to performance improvements with object lookup).